### PR TITLE
[REFACTOR][IR] Migrate Pass/PassContext ObjectRef to not-null

### PR DESCRIPTION
**Pull request by ANSHUMAN87**
Refer #5318

@tqchen, @jroesch  : Please help review, Thanks!

### CODE REVIEW

File: include/tvm/ir/transform.h [hunk 1]
```diff
@@ -161,7 +161,7 @@ class PassContextNode : public Object {
  */
 class PassContext : public ObjectRef {
  public:
-  PassContext() {}
+  PassContext() : PassContext(make_object<PassContextNode>()) {}
```
Comment (tqchen) [inline include/tvm/ir/transform.h:164]: use https://github.com/apache/incubator-tvm/blob/master/include/tvm/runtime/object.h#L712 and add  `PassContextNode* operator->()`; you will need copy constructor and other things now that default constructor is defined in a different way.

File: src/relay/backend/compile_engine.cc [hunk 2]
```diff
@@ -689,7 +689,7 @@ class CompileEngineImpl : public CompileEngineNode {
       cache_node->funcs = (*f)(cfunc->schedule, all_args, cache_node->func_name, key->source_func);
     } else {
       using tvm::transform::PassContext;
-      With<PassContext> fresh_pass_ctx_scope(PassContext::Create());
+      With<PassContext> fresh_pass_ctx_scope(static_cast<PassContext>(PassContext()));
```
Comment (tqchen) [inline src/relay/backend/compile_engine.cc:692]: I don't think static cast would be necessary here, With<PassContext> fresh_pass_ctx_scope(PassContext()); should suffice once we defined the copy constructor

Comment (ANSHUMAN87) [reply src/relay/backend/compile_engine.cc:692]: @tqchen : Thanks! I tried with copy constructor [using](https://github.com/apache/incubator-tvm/blob/9151a519b9495483a3d2bf552c8d6cf02e2302a3/include/tvm/runtime/object.h#L686), but it did not work out.

Comment (tqchen) [reply src/relay/backend/compile_engine.cc:692]: Would be interesting to see what error message you got. As copy constructor is clearly defined in there.
We can also use `With<PassContext> fresh_pass_ctx_scope();`, which invokes the default constructor(in this case the new pass context.

Comment (ANSHUMAN87) [reply src/relay/backend/compile_engine.cc:692]: @tqchen : Sorry for late reply! If i dont use static_cast [the testcase fails with additional assert](https://github.com/apache/incubator-tvm/blob/6ae439c8c58dd0118d2f2c5d1c4bcb650df47104/tests/python/relay/test_pass_manager.py#L555) count.

Comment (ANSHUMAN87) [reply src/relay/backend/compile_engine.cc]: @tqchen : I have changed it to default constructor now, please check. Thanks!

Comment (ANSHUMAN87) [reply src/relay/backend/compile_engine.cc]: > @tqchen : I have changed it to default constructor now, please check. Thanks!

@tqchen : Sorry! This change does not work. I made a mistake in code build, cause of that it got passed last time! So i have to use static_cast for the test case to pass.

Comment (ANSHUMAN87) [reply src/relay/backend/compile_engine.cc:692]: @tqchen : Any thoughts on this comment?

**Comment by tqchen** [comment]
close for now and let us revisit later. Thanks @ANSHUMAN87

**Pull request closed without merging**